internal/api/permissions.go

```diff
@@ -72,6 +90,169 @@ func (r *Router) checkAction(c echo.Context) error {
+// This is the permissions check endpoint.
+// It will return a 200 if the subject is allowed to perform all requested resource actions.
+// It will return a 400 if the request is invalid.
+// It will return a 403 if the subject is not allowed to perform all requested resource actions.
+//
+// Note that this expects a JWT token to be present in the request. This token must
+// contain the subject of the request in the "sub" claim.
+func (r *Router) checkAllActions(c echo.Context) error {
+	ctx, span := tracer.Start(c.Request().Context(), "api.checkAllActions")
+	defer span.End()
+
+	// Subject validation
+	subject, err := currentSubject(c)
+	if err != nil {
+		return echo.NewHTTPError(http.StatusBadRequest, "failed to get the subject").SetInternal(err)
+	}
+
+	subjectResource, err := r.engine.NewResourceFromID(subject)
+	if err != nil {
+		return echo.NewHTTPError(http.StatusBadRequest, "error processing subject ID").SetInternal(err)
+	}
+
+	var reqBody checkPermissionsRequest
+
+	if err := c.Bind(&reqBody); err != nil {
+		return echo.NewHTTPError(http.StatusBadRequest, "error parsing request body").SetInternal(err)
+	}
+
+	var errs []error
+
+	results := make([]*checkStatus, len(reqBody.Actions))
+
+	for i, check := range reqBody.Actions {
+		if check.Action == "" {
+			errs = append(errs, fmt.Errorf("check %d: %w", i, ErrNoActionDefined))
+
+			continue
+		}
+
+		resourceID, err := gidx.Parse(check.ResourceID)
+		if err != nil {
+			errs = append(errs, fmt.Errorf("check %d: %w: error parsing resource id: %s", i, err, check.ResourceID))
+
+			continue
+		}
+
+		resource, err := r.engine.NewResourceFromID(resourceID)
+		if err != nil {
+			errs = append(errs, fmt.Errorf("check %d: %w: error creating resource from id: %s", i, err, resourceID.String()))
+
+			continue
+		}
+
+		results[i] = &checkStatus{
+			Resource: resource,
+			Action:   check.Action,
+		}
+	}
+
+	if len(errs) != 0 {
+		return echo.NewHTTPError(http.StatusBadRequest, "invalid check request").SetInternal(multierr.Combine(errs...))
+	}
+
+	checkCh := make(chan *checkStatus)
+
+	wg := new(sync.WaitGroup)
+
+	for i := 0; i < r.concurrentChecks; i++ {
+		wg.Add(1)
+
+		go func() {
+			defer wg.Done()
+
+			for check := range checkCh {
+				// Check the permissions
+				err := r.engine.SubjectHasPermission(ctx, subjectResource, check.Action, check.Resource)
+				if err != nil {
+					check.Error = err
+				}
+			}
+		}()
+	}
+
+	wg.Add(1)
+
+	go func() {
+		defer wg.Done()
+
+		for _, check := range results {
+			checkCh <- check
+		}
+
+		close(checkCh)
+	}()
+
+	doneCh := make(chan struct{})
+
+	go func() {
+		defer close(doneCh)
+
+		wg.Wait()
+	}()
+
+	select {
+	case <-doneCh:
+	case <-ctx.Done():
+		return echo.NewHTTPError(http.StatusInternalServerError, "request cancelled").WithInternal(ctx.Err())
+	case <-time.After(maxCheckDuration):
+		return echo.NewHTTPError(http.StatusInternalServerError, "checks didn't complete in time")
```

Comment:
Setting up a context using [`context.WithTimeout`](https://pkg.go.dev/context#WithTimeout) is more idiomatic here and lets us handle all errors in one place.

Reply:
Good point, updated.